Mathlib/Algebra/Order/Ring/Cone.lean

```diff
@@ -1,68 +1,82 @@
-      rw [sub_zero]
-      exact C.mul_pos x y (by rwa [← sub_zero x]) (by rwa [← sub_zero y]) }
+We also provide constructors that convert between
+cones in rings and the corresponding ordered rings.
+-/
 
-/-- Construct a `LinearOrderedRing` by
-designating a positive cone in an existing `Ring`. -/
-def LinearOrderedRing.mkOfPositiveCone (C : TotalPositiveCone α) : LinearOrderedRing α :=
-  { LinearOrderedAddCommGroup.mkOfPositiveCone C.toTotalPositiveCone,
-    StrictOrderedRing.mkOfPositiveCone C.toPositiveCone_1 with }
+/-- `RingConeClass S R` says that `S` is a type of cones in `R`. -/
+class RingConeClass (S R : Type*) [Ring R] [SetLike S R]
+    extends AddGroupConeClass S R, SubsemiringClass S R : Prop
+
+/-- A (positive) cone in a ring is a subsemiring that
+does not contain both `a` and `-a` for any nonzero `a`.
+This is equivalent to being the set of non-negative elements of
+some order making the ring into a partially ordered ring. -/
+structure RingCone (R : Type*) [Ring R] extends Subsemiring R, AddGroupCone R
+
+/-- Interpret a cone in a ring as a cone in the underlying additive group. -/
+add_decl_doc RingCone.toAddGroupCone
+
+instance RingCone.instSetLike (R : Type*) [Ring R] : SetLike (RingCone R) R where
+  coe C := C.carrier
+  coe_injective' p q h := by cases p; cases q; congr; exact SetLike.ext' h
+
+instance RingCone.instRingConeClass (R : Type*) [Ring R] :
+    RingConeClass (RingCone R) R where
+  add_mem {C} := C.add_mem'
+  zero_mem {C} := C.zero_mem'
+  mul_mem {C} := C.mul_mem'
+  one_mem {C} := C.one_mem'
+  eq_zero_of_mem_of_neg_mem {C} := C.eq_zero_of_mem_of_neg_mem'
+
+namespace RingCone
+
+variable {T : Type*} [OrderedRing T] {a : T}
+
+variable (T) in
+/-- Construct a cone from the set of non-negative elements of a partially ordered ring. -/
+def nonneg : RingCone T where
+  __ := Subsemiring.nonneg T
+  eq_zero_of_mem_of_neg_mem' {a} := by simpa using ge_antisymm
+
+@[simp] lemma nonneg_toSubsemiring : (nonneg T).toSubsemiring = .nonneg T := rfl
+@[simp] lemma nonneg_toAddGroupCone : (nonneg T).toAddGroupCone = .nonneg T := rfl
+@[simp] lemma mem_nonneg : a ∈ nonneg T ↔ 0 ≤ a := Iff.rfl
+@[simp, norm_cast] lemma coe_nonneg : nonneg T = {x : T | 0 ≤ x} := rfl
+
+instance nonneg.isMaxCone {T : Type*} [LinearOrderedRing T] : IsMaxCone (nonneg T) where
+  mem_or_neg_mem := IsMaxCone.mem_or_neg_mem (C := AddGroupCone.nonneg T)
+
+end RingCone
+
+variable {S R : Type*} [Ring R] [SetLike S R] (C : S)
+
+/-- Construct a partially ordered ring by designating a cone in a ring. -/
+@[reducible] def OrderedRing.mkOfCone [RingConeClass S R] : OrderedRing R where
+  __ := ‹Ring R›
+  __ := OrderedAddCommGroup.mkOfCone C
+  zero_le_one := show _ ∈ C by simpa using one_mem C
+  mul_nonneg x y xnn ynn := show _ ∈ C by simpa using mul_mem xnn ynn
+
+/-- Construct a linearly ordered domain by designating a maximal cone in a domain. -/
+@[reducible] def LinearOrderedRing.mkOfCone
```

Comment:
Please add a warning saying that using this as a constructor in an instance can lead to diamonds due to non-customizable `lt`, `decidableLT`, `decidableEq`, and `compare` fields.

Reply:
Done -- does this also apply to the previous `def`?